### PR TITLE
gh-80789: Implement build-time pip bundling in ``ensurepip``

### LINKED ISSUE #80789: Replace bundled pip and setuptools with a downloader in the ensurepip module
BPO | [36608](https://bugs.python.org/issue36608)
--- | :---
Nosy | @ericvsmith, @ned-deily, @serhiy-storchaka, @dstufft, @webknjaz, @pradyunsg
PRs | <li>python/cpython#12791</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2019-04-11.22:13:02.045>
labels = ['3.8', 'type-feature', 'library', '3.9']
title = 'Replace bundled pip and setuptools with a downloader in the ensurepip module'
updated_at = <Date 2021-07-05.23:28:08.628>
user = 'https://github.com/webknjaz'
```

bugs.python.org fields:
```python
activity = <Date 2021-07-05.23:28:08.628>
actor = 'ned.deily'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Library (Lib)']
creation = <Date 2019-04-11.22:13:02.045>
creator = 'webknjaz'
dependencies = []
files = []
hgrepos = []
issue_num = 36608
keywords = ['patch']
message_count = 6.0
messages = ['339998', '340039', '340060', '340169', '340218', '347836']
nosy_count = 6.0
nosy_names = ['eric.smith', 'ned.deily', 'serhiy.storchaka', 'dstufft', 'webknjaz', 'pradyunsg']
pr_nums = ['12791']
priority = 'normal'
resolution = None
stage = 'patch review'
status = 'open'
superseder = None
type = 'enhancement'
url = 'https://bugs.python.org/issue36608'
versions = ['Python 3.8', 'Python 3.9']
```

</p></details>

### Linked PRs
* gh-109130
* gh-109245

### PR DESCRIPTION
**Pull request by webknjaz**
Prior to this patch, Pip wheels were stored in the Git repository of
CPython. Git is optimized for text but these artifacts are binary. So
the unpleasant side effect of doing this is that the bare Git
repository size is being increased by the zip archive side every time
it is added, removed or modified. It's time to put a stop to this.

The patch implements an `ensurepip.bundle` module that is meant to be
called through `runpy` to download the Pip wheel and place it into the
same location as before. It removes the wheel file from the Git
repository and prevents re-adding it by defining a new `.gitignore`
configuration file.

The idea is that the builders of CPython are supposed to invoke the
following command during the build time:

```console
$ python -m ensurepip.bundle
```

This command will verify the existing wheel's SHA-256 hash and, if it
does not match, or doesn't exist, it will proceed to download the
artifact from PyPI. It will confirm its SHA-256 hash before placing it
into the `Lib/ensurepip/_bundled/` directory.

The only change concerning the Pip maintainers submitting pip version
bumps to the CPython repo is that they will no longer need to check in
the wheel file binary, but would instead have to update SHA-256 and pip
version strings at the end of `Lib/ensurepip/_structs.py`. Previously,
the pip version used to be hard-coded at the top of
`Lib/ensurepip/__init__.py` and SHA-256 was not included.

Every single line added or modified as a part of this change is also
covered with tests. Every new module has 100% coverage. The only
uncovered lines under `Lib/ensurepip/` are the ones that are
absolutely unrelated to this effort.

Resolves #80789.

Ref: https://bugs.python.org/issue36608.

## Coverage
```console
$ ./cpython-venv/bin/python -m coverage run --pylib --source=ensurepip Lib/test/test_ensurepip.py
...................................
----------------------------------------------------------------------
Ran 35 tests in 0.048s

OK
(cpython-pyenv-py3.11.4) 

$ ./cpython-venv/bin/python -m coverage report --show-missing --skip-covered
Name                        Stmts   Miss Branch BrPart  Cover   Missing
-----------------------------------------------------------------------
Lib/ensurepip/__init__.py      79      5     28      0    93%   21-39
Lib/ensurepip/__main__.py       2      2      0      0     0%   1-2
-----------------------------------------------------------------------
TOTAL                         226      7     77      0    97%

5 files skipped due to complete coverage.
```

https://bugs.python.org/issue36608

* Issue: gh-80789

### CODE REVIEW

**Comment by webknjaz** [comment]
@ezio-melotti why did you remove the `CLA-signed` label?

**Comment by ezio-melotti** [comment]
We now switched to the CLA bot, so the labels are no longer needed/used.  I closed and reopened the PR to trigger the bot, and the CLA Signing check passed.

**Comment by webknjaz** [comment]
Thanks for the explanation!

**Comment by webknjaz** [comment]
Status update: during the EuroPython 2023 Sprints in Prague, I've started working on the conflict resolution here. There's been a number of changes to `ensurepip` since I started this PR during EuroPython 2019 in Basel, so it took me a while to regain the context and incorporate those changes into the branch. It is now in sync locally, but I'd like to ensure complete test coverage of the new code, which is why I'm delaying the push. I'm going to share that update as soon as I deal with the last bits of the uncovered code.

File: Makefile.pre.in [hunk 1]
```diff
@@ -2759,6 +2760,12 @@ patchcheck: all
 check-limited-abi: all
 	$(RUNSHARED) ./$(BUILDPYTHON) $(srcdir)/Tools/build/stable_abi.py --all $(srcdir)/Misc/stable_abi.toml
 
+.PHONY: bundle-ensurepip-dists
+bundle-ensurepip-dists:
+	if test "x$(ENSUREPIP)" != "xno"  ; then \
+		$(RUNSHARED) $(PYTHON_FOR_BUILD) -m ensurepip.bundle -v ; \
```
Comment (webknjaz) [inline Makefile.pre.in]: :exclamation: I need help with this — my attempts to integrate the bundler into the build process fail in some CI jobs (supposedly because of the `./configure` CLI flags).

**Comment by AA-Turner** [comment]
@webknjaz please avoid force-pushes, it makes reviewers' lives harder.

A

File: .github/workflows/reusable-docs.yml [hunk 2]
```diff
@@ -100,7 +100,7 @@ jobs:
     - name: 'Configure CPython'
       run: ./configure --with-pydebug
     - name: 'Build CPython'
-      run: make -j4
+      run: make all-with-ensurepip-dists-bundled -j4
```
Comment (AA-Turner) [inline .github/workflows/reusable-docs.yml:103]: Do we need ensurepip for the doctests?

File: Makefile.pre.in [hunk 3]
```diff
@@ -617,6 +617,9 @@ all:		@DEF_MAKE_ALL_RULE@
 build_all:	check-clean-src $(BUILDPYTHON) platform sharedmods \
 		gdbhooks Programs/_testembed scripts checksharedmods rundsymutil
 
+.PHONY: all-with-ensurepip-dists-bundled
+all-with-ensurepip-dists-bundled: build_all bundle-ensurepip-dists
```
Comment (AA-Turner) [inline Makefile.pre.in]: The name is quite long, and I'm unsure what 'dists' means here:
```suggestion
.PHONY: all-with-bundled-ensurepip
all-with-bundled-ensurepip: build_all bundle-ensurepip
```

File: Makefile.pre.in [hunk 4]
```diff
@@ -2759,6 +2762,14 @@ patchcheck: all
 check-limited-abi: all
 	$(RUNSHARED) ./$(BUILDPYTHON) $(srcdir)/Tools/build/stable_abi.py --all $(srcdir)/Misc/stable_abi.toml
 
+.PHONY: bundle-ensurepip-dists
+bundle-ensurepip-dists: $(BUILDPYTHON) platform
```
Comment (AA-Turner) [inline Makefile.pre.in]:
```suggestion
.PHONY: bundle-ensurepip
bundle-ensurepip: $(BUILDPYTHON) platform
```

**Review by AA-Turner — COMMENTED** [review]
Brief initial look, I haven't gone over the changes to ``ensurepip`` fully.

I did want to ask on intended use scenarios here, though -- from Pradyun & Serhiy's comments on the issue, it seemed the goal was to have the release / build process pull in the latest version of ``pip`` (similar-ish to how other dependencies are vendored).

This PR though seems to implement a new ``ensurepip.bundler`` mechanism, which exposes a new module and has unclear semantics--is this for end users, redistributors, etc? If possible, perhaps we could either make the ``_bundler`` module private, or move the bundling mechanisms to ``Tools/`` for inclusion in to the build process.

Let me know if you'd like a more detailed review into the changes in ``ensurepip`` itself, I wanted to hold off for now as of these questions on the design level.

Thanks,
Adam

**Comment by webknjaz** [comment]
> please avoid force-pushes, it makes reviewers' lives harder.

Totally. It's just that nobody was reviewing it, so I didn't bother. Then, when I was resolving merge conflicts, I messed up while trying to avoid a “foxtrot merge” and at that point, there was no other way to clean it up than to remake it. I also chose not to preserve the original carefully maintained atomic commits that were on the branch for the past 4 years, as a lot changed over time and I decided that there was no value in recovering them.

Going forward, I'll try to remember adding new commits on top. I don't always do this, so this is not my default habit. Thanks for the suggestion!

File: .github/workflows/reusable-docs.yml (hunk 2 above)
Comment (webknjaz) [reply .github/workflows/reusable-docs.yml:103]: Not sure. I thought something was failing, but that was at the time when I was trying to figure out the Makefile integration, and it didn't work. So maybe not, this needs some experimentation.

**Comment by webknjaz** [comment]
> I did want to ask on intended use scenarios here, though -- from Pradyun & Serhiy's comments on the issue, it seemed the goal was to have the release / build process pull in the latest version of `pip` (similar-ish to how other dependencies are vendored).

Yes. Though, maybe not the latest but pinned though a constant. The main idea is that by stopping updating binaries, the bare Git repo would not grow as dramatically on each pip version bump.

I'm actually expecting a review from @pradyunsg at least. And maybe from someone who would have to deal with this on the build/release process side.

File: Makefile.pre.in (hunk 3 above)
Comment (webknjaz) [reply Makefile.pre.in]: > I'm unsure what 'dists' means

Basically, this refers to Python distribution packages. As in wheel zipfiles or sdist tarballs.
Historically, `ensurepip` shipped with two wheels — `pip` and `setuptools`. Some time ago, Pradyun removed the `setuptools` one, but the infrastructure for having an iterable of dists remained in the source code. Hence the plural form.

Also, this make target doesn't control whether the ensurepip module is included by itself. It only downloads the wheel into the in-module location. So I think implying that this bundles ensurepip in the make target name is inaccurate.

When Pradyun and I last discussed it, we agreed that the list of things is probably no longer needed. So I was going to have a follow-up PR to replace it with a singular var at some point. I didn't want to include refactoring in this patch — it would make it harder to manage a non-atomic PR. But I suppose that follow-up could change the target name to specifically say “pip” instead of “dists”.

FWIW, I had problems embedding this into the Makefile. So maybe the RMs would have other preferences.

File: Makefile.pre.in (hunk 4 above)
Comment (webknjaz) [reply Makefile.pre.in]: See https://github.com/python/cpython/pull/12791#discussion_r1307351594 for why this feels incorrect.

**Comment by webknjaz** [comment]
> This PR though seems to implement a new `ensurepip.bundler` mechanism, which exposes a new module and has unclear semantics--is this for end users, redistributors, etc? If possible, perhaps we could either make the `_bundler` module private, or move the bundling mechanisms to `Tools/` for inclusion in to the build process.

That's a fair point. This is intended for RMs (and redistributors, should they build from Git and not the published tarball I suppose). Though, exposing this to the end-users might be a way to enable them to verify the integrity of their `ensurepip` copy — the bundler would only make the download if the SHA-256 hash doesn't match.
I think that making it private is definitely an option. Would love to hear from @pradyunsg regarding this.
As for moving under `Tools/`, I don't think this is a good idea, since there's no proper mechanism for running tests against them AFAIK. And I really wanted this to be well-tested. Maybe we could make an entry-point there that would still call `ensurepip._bundle` but then, why not just call it directly. The logic is kinda coupled with `ensurepip` on some level (like a shared pinned pip version and data structures).

File: .github/workflows/reusable-docs.yml [hunk 5]
```diff
@@ -100,7 +100,7 @@ jobs:
...
     - name: 'Install build dependencies'
       run: make -C Doc/ PYTHON=../python venv
```
Comment (webknjaz) [inline .github/workflows/reusable-docs.yml:105]: @AA-Turner so I looked into whether bundling pip into ensurepip is needed in doctests and the answer is “yes”. This line calls `../python -m venv` (so it uses the Python build under test and not some OS-provided copy). The `venv` module relies on `ensurepip` to provision its `site-packages/`. When there's no pip wheel, it crashes (as it should!).

**Comment by pfmoore** [comment]
I haven't had time to review the code fully (yet) but I do have a high level question. What is the intended workflow for a pip release once this change has been made? Currently, when we release a new version of pip, we create a PR for CPython that updates the bundled pip version. That PR is backported to all supported versions of CPython (where "supported" means both "still accepting PRs" and "supported by the new pip version").

With the new approach, what will we need to do when we release a new version of pip, if anything?

Some specific scenarios:

1. If we release a new pip version, but want to hold off on putting it into CPython until we're sure we won't need a bugfix release, how do we do that? This would only be a problem if a pip release and a CPython release were happening close together, but it *is* possible - pip 23.2 and Python 3.12rc1 were a good example.
2. If we drop support in pip for a Python version, how do we ensure that pip version won't be included in the desupported Python version? This shouldn't be an issue, as we would normally not desupport a Python version until it's no longer supported by the core team, but theoretically it could happen.

Basically, these boil down to - can we override the default of "just take the latest pip version", and how do we communicate with the Python release team on this?

**Comment by webknjaz** [comment]
> Basically, these boil down to - can we override the default of "just take the latest pip version", and how do we communicate with the Python release team on this?

@pfmoore There's no such default, and I haven't changed it. @AA-Turner was mistaken in his observation. There's still a hard-coded pip version in the source code: https://github.com/python/cpython/pull/12791/files#diff-f5e321563d5bc98371a960662f654a9387fd24f9bacb382d5b90d417ad2569c7R243-R244. The new thing is that a binary in Git is replaced by a SHA-256 hash. Whenever the new bundler is invoked, it checks the integrity of the existing (gitignored) pip wheel on disk. If the checksum doesn't match or the artifact is absent, it downloads it from PyPI by that hardcoded version, and checks it against that hard-coded SHA-256 before finally writing it on disk.
The runtime `ensurepip` behavior remains unchanged.

> If we drop support in pip for a Python version, how do we ensure that pip version won't be included in the desupported Python version?

That's an interesting question. I suppose it is currently decided by a human. That hasn't changed. But I can imagine that there could be an automation / test that could check the metadata in the wheel too. This is not intended to be changed in the PR, but certainly is an appealing future improvement aspect.

> Currently, when we release a new version of pip, we create a PR for CPython that updates the bundled pip version. That PR is backported to all supported versions of CPython (where "supported" means both "still accepting PRs" and "supported by the new pip version").
>
> With the new approach, what will we need to do when we release a new version of pip, if anything?

The old process involved updating the hard-coded pip version and committing the actual wheel file. The new process is to update the hard-coded pip version and the corresponding SHA-256 hash that can be copied from PyPI, for example, or calculated locally. There are no other changes concerning the pip maintainers in this workflow — the patch is smaller, the backports are the same.

File: Lib/ensurepip/_structs.py [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+
+    def as_pathlib_ctx(self) -> AbstractContextManager[Path]:
+        """Make a context manager exposing a :mod:`pathlib` instance."""
+        return nullcontext(self.wheel_path)
+
+
+PIP_REMOTE_DIST = _RemoteDistributionPackage(
+    'pip',
+    '23.2.1',
+    '7ccf472345f20d35bdc9d1841ff5f313260c2c33fe417f48c30ac46cccabf5be',
```
Comment (webknjaz) [inline Lib/ensurepip/_structs.py:244]: @pfmoore this is where the version information has moved to. I was considering moving it into a config file (like toml?) but that's something I wouldn't want to couple with this PR, it's big as it is.

File: PCbuild/build.bat [hunk 7]
```diff
@@ -175,6 +175,15 @@ echo on
...
  %1 %2 %3 %4 %5 %6 %7 %8 %9
 
+if "%platf%"=="Win32" set prefix=%dir%win32
+if "%platf%"=="x64" set prefix=%dir%amd64
+if "%platf%"=="ARM64" set prefix=%dir%arm64
+if "%platf%"=="ARM" set prefix=%dir%arm32
+set suffix=
+if "%conf%"=="Debug" set suffix=_d
+set built_python_exe=%prefix%\python%suffix%.exe
+call "%built_python_exe%" -m ensurepip.bundle -v
```
Comment (webknjaz) [inline PCbuild/build.bat]: I know this is ugly, but I don't know much about Windows :(
Improvement suggestions are welcome!

Comment (zware) [inline PCbuild/build.bat]: There's a handy `"%dir%\..\python.bat"` that you can use instead.

File: Makefile.pre.in (hunk 4 above)
Comment (zware) [inline Makefile.pre.in]: ISTM this target name should explicitly call out that it is (probably) downloading something from the internet, something more along the lines of `download-ensurepip-blobs`.

**Review by zware — COMMENTED** [review]
I am very much on board with the stated goal of this PR of removing the binary blobs from the repository.  I'm not fully convinced by the approach, though: this adds quite a lot of code to download (currently, and probably never again more than) one blob, and it took me several minutes of searching to find where the actual download was happening, which turns out to be buried in a module called `_structs`.

ISTM that if we're reaching out to the internet anyway, we could have a much lower maintenance burden here by making use of `bootstrap.pypa.io/get-pip.py` which could easily grow a `--download-only` option with a ~3 line implementation.  `get-pip.py --no-setuptools --no-wheel --download-only 'pip=={pip_version}' -w Lib/ensurepip/_bundled` would pretty much do everything we want as far as I can tell.  If we used a requirements file with it instead of the versioning argument, we could even have hash verification.

**Comment by webknjaz** [comment]
_Note to myself:_ this will need a companion change @ https://github.com/python/buildmaster-config/blob/6fed7ed/master/custom/factories.py#L79.

File: PCbuild/build.bat (hunk 7 above)
Comment (webknjaz) [reply PCbuild/build.bat]: Is it guaranteed to locate the built interpreter version? I was a bit lost and haven't found how this file is generated to check it myself..

File: Makefile.pre.in (hunk 4 above)
Comment (webknjaz) [reply Makefile.pre.in]: Fair point. I like this suggestion over https://github.com/python/cpython/pull/12791#discussion_r1306868995.

File: PCbuild/build.bat (hunk 7 above)
Comment (zware) [reply PCbuild/build.bat]: Yes; see https://github.com/python/cpython/blob/4116592b6f014a2720e9b09e2c8dec4bf4b4cd8f/PCbuild/python.vcxproj#L134-L150

(Note that `%dir%\..\python.bat` and `%dir%\find_python.bat` are very different :).  You want the former, which is also used in the PGO section several lines up.)

File: Makefile.pre.in (hunk 4 above)
Comment (webknjaz) [reply Makefile.pre.in]: I've applied this change, so I'm marking the thread as resolved.

File: Makefile.pre.in (hunk 3 above)
Comment (webknjaz) [reply Makefile.pre.in]: I chose to make use of another make target name suggestion. See: https://github.com/python/cpython/pull/12791#discussion_r1307596457. Marking this thread as resolved.

File: Makefile.pre.in (hunk 4 above)
Comment (webknjaz) [reply Makefile.pre.in]: I chose to make use of another make target name suggestion. See: https://github.com/python/cpython/pull/12791#discussion_r1307596457. Marking this thread as resolved.

File: PCbuild/build.bat (hunk 7 above)
Comment (webknjaz) [reply PCbuild/build.bat]: I applied the suggested change. Marking this thread as resolved.

**Comment by webknjaz** [comment]
> ISTM that if we're reaching out to the internet anyway, we could have a much lower maintenance burden here by making use of `bootstrap.pypa.io/get-pip.py` which could easily grow a `--download-only` option with a ~3 line implementation. `get-pip.py --no-setuptools --no-wheel --download-only 'pip=={pip_version}' -w Lib/ensurepip/_bundled` would pretty much do everything we want as far as I can tell. If we used a requirements file with it instead of the versioning argument, we could even have hash verification.

I don't know if we should bring externally hosted executable code into the play, as it is an additional security concern to watch out for.
I'd like to hear what @pradyunsg thinks about this.

File: Lib/ensurepip/_structs.py (hunk 6 above)
Comment (pfmoore) [reply Lib/ensurepip/_structs.py:244]: Thanks. So the pip release manager would need to calculate that sha256 value? That should be OK, although while I have tools on Windows to do this, I don't know if everyone does (I assume Unix has a command for this, although I've never checked).

(Actually, I just saw that you mentioned it can be copied from PyPI, so that's sufficient).

**Comment by pfmoore** [comment]
+1 to both the points @zooba made.

> this adds quite a lot of code to download (currently, and probably never again more than) one blob, and it took me several minutes of searching to find where the actual download was happening, which turns out to be buried in a module called _structs.

I do agree with this point, though. One of the reasons I haven't gone through the code in this PR yet is it does seem pretty complex for something that's essentially just downloading a single file.

Why not just hard-code the pip version, sha, and URL (all of which can be copied from the PyPI download page) and just do a simple HTTP download? It adds a little more work for the pip RM to get the URL, but I'm OK with that (I'm speaking as a pip RM here, in case it's not obvious!) As far as I know, PyPI download URLs are permanent, so there's no risk of the location moving.

> The old process involved updating the hard-coded pip version and committing the actual wheel file. The new process is to update the hard-coded pip version and the corresponding SHA-256 hash that can be copied from PyPI, for example, or calculated locally. There are no other changes concerning the pip maintainers in this workflow — the patch is smaller, the backports are the same.

OK, so management of pip releases remains identical, which is great, as it means our existing, working, workflow can continue and we don't have to worry about edge cases like I started to do above 🙂

File: Lib/ensurepip/_structs.py (hunk 6 above)
Comment (AA-Turner) [reply Lib/ensurepip/_structs.py:244]: ``sha256sum`` is usable on Unix (& WSL), or ``openssl sha256`` on both, or ``hashlib.sha256``!

File: PCbuild/build.bat (hunk 7 above)
Comment (webknjaz) [reply PCbuild/build.bat]: @zware could you take a look at https://dev.azure.com/Python/cpython/_build/results?buildId=136036&view=logs&j=91c152bd-7320-5194-b252-1404e56e2478&t=c7e99cd8-4756-5292-d34b-246ff5fc615f&l=764 — any idea why it might be failing? I don't see how it might be related to the PR, but OTOH, it doesn't crash on `main`…

**Comment by webknjaz** [comment]
> Why not just hard-code the pip version, sha, and URL (all of which can be copied from the PyPI download page) and just do a simple HTTP download? It adds a little more work for the pip RM to get the URL, but I'm OK with that (I'm speaking as a pip RM here, in case it's not obvious!) As far as I know, PyPI download URLs are permanent, so there's no risk of the location moving.

_Fun fact:_ the very first implementation I had in this PR years ago did just that — it actually used URLs with SHA-256 in the fragment. See: https://github.com/python/cpython/commit/e2dcb69a5a1eecf8e2719d460bc4dd5cc0179440#diff-ecfb0d1f34501dd30722053b9b39dea9139ffdd30dfcdc0964bcb18b95e443e4R16-R19.
But when I got to resurrecting the effort this year, I figured that it was a bit fragile and would probably be a bit annoying to the bump PR authors. So I shifted towards a proper data structure where only version and SHA-256 would need to be provided and everything else is derived from those. This also allowed me to make use of the predictable URLs documented at https://warehouse.pypa.io/api-reference/integration-guide.html#predictable-urls: https://github.com/python/cpython/commit/c31bdca823eab7e4ffcd70c7bd9a9895506e1e33#diff-f5e321563d5bc98371a960662f654a9387fd24f9bacb382d5b90d417ad2569c7R79-R81.

**Comment by webknjaz** [comment]
> does seem pretty complex for something that's essentially just downloading a single file.

Some of it is a byproduct of trying to have a shared place for data that's used by the downloader and the normal runtime. I also wanted things to be testable and to be able to produce 100% coverage I figured that import-time evaluation present previously in some places was making it hard to achieve, which is why I changed it to push as much as possible from import- to run-time. Admittedly, this might've resulted in some overengineering.

@pfmoore would it make sense to include getting rid of the `REMOTE_DIST_PKGS` tuple in favor of a single variable to make it simpler?

**Comment by pfmoore** [comment]
I should note that I still haven't gone through the code in the existing PR in detail, so I'm not quite sure what you mean when you say "a shared place for data that's used by the downloader and the normal runtime". I'm not sure of the benefit, as the only data the downloader needs is the pip version and the hash - are you thinking of the wheel name itself? If so, then I think you're overcomplicating again - pip wheels are always `pip-{version}-py3-none-any.whl`.

File: PCbuild/build.bat (hunk 7 above)
Comment (zware) [reply PCbuild/build.bat]: That particular build is apparently trying to run an `arm64` build on what is apparently an `amd64` runner, which isn't going to work out well :)

**Comment by webknjaz** [comment]
@pfmoore that's what's implemented @ https://github.com/python/cpython/commit/c31bdca823eab7e4ffcd70c7bd9a9895506e1e33#diff-f5e321563d5bc98371a960662f654a9387fd24f9bacb382d5b90d417ad2569c7R33-R34.

Note that due to the fact that the original code has a tuple of wheels (used to be pip + setuptools) and the removal of setuptools didn't change that, this tuple remains in the code base (as I felt like that was out of the scope here + this PR was also started at the time when there were two things in that tuple).

That essentially dictates a good chunk of loops in the code base and strings like `pip` not being hardcoded. I'm sure that some of the complexity you're seeing is due to the fact that we have to iterate over that structure, even though there's just one element nowadays — pip.
I talked to Pradyun earlier, and we agreed that this needs to change, as it's unlikely that there will ever be an additional wheel that's not pip. Though, we were talking about having done that removal separately, as it doesn't feel right to stick refactoring into a feature PR. Which is why I was thinking of having this in a follow-up.

I know for sure that removing this will make it more straightforward to deal with the control flow. What's your take on this? Should I try implementing that removal prior to this PR?

**Comment by webknjaz** [comment]
@pfmoore also note that the runtime use of the wheels is dependent on a compile-time setting `WHEEL_PKG_DIR` that the implementation on `main` discovers during import time… Since I wanted to be able to test this and other things that depend on the import-time discovery (the discovered wheels are also set during import-time), I had to do something about this to avoid having to monkey-patch a million things in tests.
Maybe, this deserves a separate refactoring PR, though.

**Comment by pfmoore** [comment]
OK, I agree with @pradyunsg that we should remove all of the code that allows for multiple wheels, and strip everything back to the core job of managing a single pip wheel. If the complexity here is because you're trying to preserve generality because of the possibility of multiple wheels, then I think that's unnecessary.

I personally don't think there's any need to worry about having multiple PRs reflecting refactoring vs new features here. If you want to do an initial PR that switches to only caring about a pip wheel, and follow up with the download-at-build-time PR, then fair enough, but personally I'd just do it as a single PR.

**Comment by AA-Turner** [comment]
> As for moving under `Tools/`, I don't think this is a good idea, since there's no proper mechanism for running tests against them AFAIK.

You can look at ``test_clinic`` for inspiration; the key part is:

```python
from test import test_tools

test_tools.skip_if_missing('clinic')
with test_tools.imports_under_tool('clinic'):
    import clinic
```

A

**Comment by AA-Turner** [comment]
I agree with Paul -- you could potentially go even further and hard-code ``f'https://files.pythonhosted.org/packages/py3/p/pip/pip-{_PIP_VERSION}-py3-none-any.whl'``, though that has some drawbacks.

If keeping the bootstrapping tool in ``Lib/ensurepip/``, I think we should either make the module private, or add some clearer guidance on when / by whom it should be used.

A

**Comment by pradyunsg** [comment]
I'll echo the general sentiment that this has a lot of complexity being introduced and it would be good to reduce complexity if we're reworking how things work -- many of the changes make more sense in the context of "setuptools" and "pip" being needed and this _definitely_ does not need a public `ensurepip.bundle` module name.

> the very first implementation I had in this PR years ago did just that — it actually used URLs with SHA-256 in the fragment. See: [e2dcb69#diff-ecfb0d1f34501dd30722053b9b39dea9139ffdd30dfcdc0964bcb18b95e443e4R16-R19](https://github.com/python/cpython/commit/e2dcb69a5a1eecf8e2719d460bc4dd5cc0179440#diff-ecfb0d1f34501dd30722053b9b39dea9139ffdd30dfcdc0964bcb18b95e443e4R16-R19).
> But when I got to resurrecting the effort this year, I figured that it was a bit fragile and would probably be a bit annoying to the bump PR authors.

TBH, that's the approach I like best among everything that's come up in the discussion. I don't quite follow why you think that's fragile. And, as someone who would need to do these bump PRs, it isn't particularly annoying either (copy the URL for the relevant wheel from https://pypi.org/simple/pip/ and paste it in).

File: Lib/ensurepip/_structs.py [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+        """
+        wheel_file_name_prefix = f'{dist_pkg_name !s}-'
+
+        wheel_pkg_dir = _get_wheel_pkg_dir_from_sysconfig()
+        dist_matching_wheels = wheel_pkg_dir.glob(
+            f'{wheel_file_name_prefix !s}*.whl',
+        )
+
+        try:
+            first_matching_dist_wheel = sorted(dist_matching_wheels)[0]
```
Comment (webknjaz) [inline Lib/ensurepip/_structs.py:205]: Self note: the original code actually goes for the last matching:
```suggestion
            last_matching_dist_wheel = sorted(dist_matching_wheels)[-1]
```

**Comment by webknjaz** [comment]
So I got around to removing the looping: https://github.com/python/cpython/pull/109245.

**Comment by webknjaz** [comment]
@pradyunsg communicated IRL that it's been decided to postpone the effort indefinitely due to the contributor experience concerns. Which is why I'm closing the PR for now. If something changes, it can then be resurrected once again.
